### ISSUE TITLE
Inconsistency in standalone_eval.py and potential indexing bug in main evaluation

### ISSUE THREAD

**Issue by sbuttau** [facebookresearch/univlg#15]
Hello, I am testing the model on `scene0329_00` from ScanRefer and noticed that `standalone_eval.py` yields different Top-1 predictions compared to the full evaluation pipeline (with `VIS_REF=True`) for the same data sample.

(STANDALONE EVAL: top-1 prediction in red)
<img width="700" height="500" alt="Image" src="https://github.com/user-attachments/assets/929a4e47-e438-44c2-af24-be9c90a8c3d1" />

(FULL EVAL: top-1 prediction in red, ground truth in green)
<img width="700" height="500" alt="Image" src="https://github.com/user-attachments/assets/9af8d937-72a9-4817-a781-2617b87ed32f" />

Despite my efforts to synchronize the inputs, the results still do not match:

- Standalone Eval: The Top-1 prediction (red box) correctly identifies the target.
- Full Eval Pipeline: The Top-1 prediction (red) misses the Ground Truth (green)

Since the standalone script explicitly asserts `USE_SEGMENTS=False`, while the full pipeline uses them, could this be the primary cause of the discrepancy? Does the model's ranking logic rely heavily on the geometric consistency provided by scannet_segments, or should the mask_decoder theoretically produce identical logits regardless?

**Potential Bug** 
While investigating, I noticed a potential logic error in how voxelized metadata is handled in the main pipeline. When `USE_GHOST_POINTS` is active, `scannet_p2v_` and `max_valid_points` still seem to refer to the original raw point cloud size (~300k) instead of the voxelized size (~242k).

In the full eval pipeline ([video_mask_former_model.py](https://github.com/facebookresearch/univlg/blob/f86f0d5fa13864a518d7634e6fe3ad66f50d9f07/univlg/video_maskformer_model.py#L876)):

```
if self.cfg.USE_GHOST_POINTS and decoder_3d:

    scannet_pc_ = scatter_mean(scannet_pc, scannet_p2v, dim=1)

    scannet_p2v_ = (
        torch.arange(scannet_pc.shape[1], ...) # <--- Uses raw point count (300k)
        .unsqueeze(0).repeat(scannet_pc.shape[0], 1)
    )
[...]

outputs = self.mask_decoder(
            mask_features,
            shape=[bs, v],
            mask_features_xyz=scannet_pc_,
            mask_features_p2v=scannet_p2v_,
            segments=scannet_segments_batched if self.cfg.USE_GHOST_POINTS else segments,
            decoder_3d=decoder_3d,
            captions=captions,
            actual_decoder_3d=actual_decoder_3d,
            scannet_all_masks_batched=scannet_all_masks_batched,
            max_valid_points=[targets[i]['max_valid_points'] for i in range(len(targets))] if self.cfg.USE_GHOST_POINTS and decoder_3d else None,
            tokenized_answer=padded_answers if (self.cfg.GENERATION and self.training) else None,
            answers=answers if (self.cfg.GENERATION and self.training) else None,
        )
```

Since `scannet_pc_` has a reduced number of rows, using an arange based on the original `scannet_pc.shape[1] `creates indices that exceed the actual tensor dimensions. 

I attempted to fix this in the standalone script by synchronizing the shapes:

```
scannet_pc_ = scatter_mean(scannet_pc, scannet_p2v, dim=1)
scannet_p2v_ = (
        torch.arange(scannet_pc_.shape[1], device=scannet_pc.device)
        .unsqueeze(0)
        .repeat(scannet_pc_.shape[0], 1)
    )
max_valid_points_voxel = [scannet_pc_.shape[1]] * bs 

outputs = model.mask_decoder(
        mask_features, # torch.Size([2, 256, 38882, 1])
        shape=[bs, v],
        mask_features_xyz=scannet_pc_,
        mask_features_p2v=scannet_p2v_,
        segments=None,
        decoder_3d=True,
        captions=captions,
        actual_decoder_3d=True,
        scannet_all_masks_batched=None,
        max_valid_points=max_valid_points_voxel,
        tokenized_answer=None,
    )
```

Even with this fix, the results between the standalone and the full pipeline still diverge.
Could you confirm if `max_valid_points` should indeed be synchronized with the voxelized shape to ensure correct cross-attention? And more importantly, is it expected that the absence of segments in the standalone script leads to different query rankings, or should I look for other discrepancies in the coordinate alignment or preprocessing?

Thank you in advance!

**Comment by ayushjain1144** [comment]
Hi, 

Thank you for your detailed look into our code. Below are some answers:

> Since the standalone script explicitly asserts USE_SEGMENTS=False, while the full pipeline uses them, could this be the primary cause of the discrepancy? Does the model's ranking logic rely heavily on the geometric consistency provided by scannet_segments, or should the mask_decoder theoretically produce identical logits regardless?

USE_SEGMENTS vs not will definitely change the logits. usually, the mAP numbers are slightly higher when using segments vs not using them. but that is mostly related to crispness of the mask, rather than shifting the entire instance. Maybe the example you are seeing is a one-off instance where the model is confused between the two tables. on an average, the difference between numbers between segments vs not is not more than 1-2%

>  I noticed a potential logic error in how voxelized metadata is handled in the main pipeline.

yes, you are right. and i think it works in the main pipeline because instead of using these voxelized p2v, we use segments there. your fix is correct. [without this fix, you should even see a runtime error i think]

> max_valid_points
that is only used for box loss computation, it shouldn't be relevant to eval

Overall, the standalone eval script isn't a well tested piece of code. we wrote it very last minute and didn't check in detail. so it can contain additional bugs too. if you can, using the main pipeline code will be safer
